asset/file_reader/slicer.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+const { Slicer } = require('@terascope/job-components');
+const Queue = require('@terascope/queue');
+const { getOffsets } = require('@terascope/chunked-file-reader');
+const { TSError } = require('@terascope/utils');
+const Promise = require('bluebird');
+const path = require('path');
+const fse = require('fs-extra');
+
+class FileSlicer extends Slicer {
+    constructor(context, opConfig, executionConfig) {
+        super(context, opConfig, executionConfig);
+        this._queue = new Queue();
+        this._doneSlicing = false;
+    }
+
+    async initialize() {
+        this.checkProvidedPath();
+        this.getFilePaths(this.opConfig.path)
+            .then(() => {
+                this._doneSlicing = true;
+            })
+            .catch((err) => {
+                // This will catch an error if there is  a problem getting the initial directory
+                // contents
+                throw new Error(
+                    `There was a problem slicing files from ${this.opConfig.path}: ${err}`
+                );
+            });
+    }
+
+    checkProvidedPath() {
+        try {
+            const dirStats = fse.lstatSync(this.opConfig.path);
+            if (dirStats.isSymbolicLink()) {
+                throw new Error('Directory for `file_reader` cannot be a symlink!');
+            }
+            const dirContents = fse.readdirSync(this.opConfig.path);
+            if (dirContents.length === 0) {
+                throw new Error('Must provide a non-empty directory for `file_reader`!!');
+            }
+        } catch (err) {
+            throw new Error(`Path must be valid! Encountered the following error:\n${err}`);
+        }
+    }
+
+    async slice() {
+        // Grab a record if there is one ready in the queue
+        if (this._queue.size() > 0) return this._queue.dequeue();
+
+        // Finish slicer if the queue is empty and it's done prepping slices
+        if (this._doneSlicing) return null;
+
+        // If the queue is empty and there are still slices, wait for a new slice
+        return new Promise((resolve) => {
+            const intervalId = setInterval(() => {
+                if (this._queue.size() > 0) {
```

Comment:
If slicer finishes while this interval is going, it will never resolve, if `this._doneSlicing === true` the interval should be canceled.

Reply:
We also need to check for the `this._shutdown` to clear the interval

Reply:
fixed